src/a2rl/_dataframe.py

```diff
@@ -677,6 +677,112 @@ def add_value(
+                `SARSA <https://en.wikipedia.org/wiki/State-action-reward-state-action>`_. Must be
+                be within 0 and 1.
+            gamma: Discount factor of future reward in Q-Learning and SARSA. Must be within 0 and 1.
+            value_col: The column name for the computed values.
+            override: What to do when this dataframe has had column ``value_col``. Valid values
+                are ``replace`` to silently override, ``warn`` to show a warning, and ``raise`` to
+                raise a :exc:`ValueError`.
+            episode_identifier: group-by key in the this dataframe. Ensure that breaks BETWEEN
+                episodes are tagged with a ``0`` group name.
+
+        Returns:
+            This dataframe, modified with an additional ``value_col`` column. This return value is
+            provided to facilitate chaining as-per the functional programming style.
+        """
+        self._check_add_value_args(value_col, override, alpha, gamma)
+        if len(self._rewards) == 1:
+            df = self
+        else:
+            df = WiDataFrame(
+                self,
+                states=self.states,
+                actions=self.actions,
+                rewards=self.rewards[:1],
+            )
+        df_t = wi.DiscreteTokenizer(n_bins=50).fit_transform(df.trim())
+
+        def calc_q_values(sub, df, x, alpha, gamma, sarsa=sarsa):
+            # print(x.name)
+            if sub.name == 0:
+                return pd.Series(0, index=sub.index)
+
+            df = df.loc[sub.index]
+            x = x.loc[sub.index]
+
+            # Temp df with only three columns: _state, _action, _reward
+            df_t = pd.concat(  # type: ignore[assignment]
+                [
+                    x[x.states].astype(str).apply("_".join, axis=1).astype("category").cat.codes,
+                    x[x.actions].astype(str).apply("_".join, axis=1).astype("category").cat.codes,
+                    df[df.rewards],  # .reset_index(drop=False),
+                ],
+                axis=1,
+                copy=False,
+            )
+
+            df_t.columns = ["_state", "_action", "_reward"]  # type: ignore[assignment]
+
+            q_table = np.zeros([df_t["_state"].nunique(), df_t["_action"].nunique()])
+
+            iterations = 10
+
+            for n in range(iterations):
+                for i in range(df_t.index[0], df_t.index[-1]):
+                    state = int(df_t.loc[i, "_state"])
+                    next_state = int(df_t.loc[i + 1, "_state"])
+                    action = int(df_t.loc[i, "_action"])
+                    reward = df_t.loc[i + 1, "_reward"]
+                    old_value = q_table[state, action]
+
+                    if sarsa:
+                        next_value = q_table[next_state, np.argmax(q_table[next_state])]
+                        new_value = old_value + alpha * (reward + gamma * next_value - old_value)
```

Comment:
Remove the (1-alpha) term from `old_value`. Aligned with canonical q-value or sarsa formula, and in-line with PR #21 which fixes `add_value()`.